### DISCUSSION TITLE
How to output the nodal multi-variable value as CSV file?

### DISCUSSION THREAD

**Discussion by Yulongggggg** [idaholab/moose#29099] [Q&A Modules: General] [answered]
Hi developers, recently, I have been trying to use MOOSE to output all the variables on all the nodes or elements. I want a CSV with every row with the x and y coordinates of every node and corresponding variables (dispx, dispy, stress_xx, material properties etc.), but in MOOSE the displacement is nodal value and the stress, strain and material properties are element value, so if there's a way to change the element value to nodal value or change nodal value to element value so they can have be present in the same CSV file
`  [Nodalvalue]
    type = NodalValueSampler
    sort_by = 'x'
    variable = 'disp_x disp_y'
    execute_on = 'TIMESTEP_END'
  []
  [Elemvalue]
    type = ElementValueSampler
    sort_by = 'x'
    variable = 'vonmises_stress stress_xx stress_xy stress_yy strain_xx strain_xy strain_yy'
    execute_on = 'TIMESTEP_END'
  []`

**Comment by GiudGiud** [comment]
Hello

You can use the ProjectionAux to project to or from a nodal variable. 
This will be good for postprocessing but there will be a projection error in some cases, which you can evaluate using the L2 norm different postprocessors

Guillaume

**Reply by Yulongggggg** [reply]
Dear Guid,

Thank you so much for your prompt reply. I have checked the ProjectionAux and examples but still can't understand how to use it. 

(1) After my simulation, I will get a 'nodal_value file' & 'element_value file,' which are assigned to nodes and Gaussian points. what I want to achieve is to unify all the points (i.e., make all the variables in the nodes rather than Gaussian points, like stress & strain), this is my current code:
`[VectorPostprocessors]
  [Nodalvalue]
    type = NodalValueSampler
    sort_by = 'x'
    variable = 'disp_x disp_y'
    execute_on = 'TIMESTEP_END'
  []
  [Elemvalue]
    type = ElementValueSampler
    sort_by = 'x'
    variable = 'vonmises_stress stress_xx stress_xy stress_yy strain_xx strain_xy strain_yy'
    execute_on = 'TIMESTEP_END'
  []`

(2) And can you teach me a little bit about how to assign the corresponding material properties to the points, I have tried using but still don't know how to play with it. Thank you so much, and this is the link to my whole code"https://github.com/Yulongggggg/test"
' [materialvalue]
   type = ElementMaterialSampler
   material = Materials
 []'

**Reply by GiudGiud** [reply]
Hello

The ElementMaterialSampler should be sampling at the quadrature points. Are you asking about how to sample at nodes for material properties?

Can you unify at the sampler level ? E.g. sample all the variables using a NodalValueSampler or an ElementValueSampler

Guillaume

**Reply by Yulongggggg** [reply]
Hi, I have tried. If i implement like the image, it will pose error:"(VectorPostprocessors/Elemvalue/variable):check_inp(moose_srv)
    The variable 'disp_x' is not elemental.check_inp(moose_srv)"
<img width="824" alt="Screenshot 2024-11-20 at 10 59 19 AM" src="https://github.com/user-attachments/assets/a818e6ce-78a1-472f-bfb4-bd171727f0ed">

**Reply by Yulongggggg** [reply]
Q: for (The ElementMaterialSampler should be sampling at the quadrature points. Are you asking about how to sample at nodes for material properties?)

A: Is there a way I can sample all the variables at only nodes or only quadrature points? The reason I want to do it is I want put them into Machine learning modes, so I wish each point (no matter nodes or quadrature points) can contain all informations like disp, stress, strain

**Reply by Yulongggggg** [reply]
> ####

what I mean is I want to implement it in the post-process part

**Reply by Yulongggggg** [reply]
@GiudGiud

**Reply by GiudGiud** [reply]
```
The variable 'disp_x' is not elemental.check_inp(moose_srv)"
```
let's just comment this out for now I think it will work for nodal variables.
Ill push a patch to enable this soon

**Reply by Yulongggggg** [reply]
> this

Thank you! I wish there could be all the information on the quadrature points, for now, the displacements only exist on the nodal points, and it would be perfect to get displacements also on the quadrature points. When you enable the function could you please tell me? Thank you

**Reply by GiudGiud** [reply]
You can already use these objects instead btw:
QuadraturePointsPositions to gather the quadrature point positions (in the Positions block)
PositionsFunctorValueSampler to sample all the variables (using their functor base class)

**Reply by GiudGiud** [reply]
I looked and the ElementValueSampler will never be able to output nodal variables.
I recommend you use the PositionsFunctorValueSampler option.

**Reply by Yulongggggg** [reply]
Hi Giud, I still don't know how to implement this, now I have two CSV files. (1) The element_value file has the coordinate of every quadrature point with variable values like stress and strain but no displacement value. (2) The Nodal_value file has the coordinate of every nodal point but only has the displacement value, is there a way to give every quadrature point the displacement value?

**Reply by GiudGiud** [reply, marked as answer]
If you do what I mention here, you'll be down to 1 CSV file
add this:

```
[Positions]
  [all_mesh]
    type = QuadraturePointsPositions
  []
```
and this
```
[VectorPostprocessors]
  [point_sample]
    type = PositionsFunctorValueSampler
    functors = 'disp_x disp_y disp_z stress strain'
    positions = 'all_mesh'
    execute_on = TIMESTEP_END
  []
[]
```

**Reply by Yulongggggg** [reply]
Thank you so much Guid, I can't image how hard it could be without your help. I used the 'ElementCentroidPositions' instead so I can get the value from every element centroid points. Now I have all the variables just like the image. Just to make sure, there's no way to do the same to all the nodal points, right? Then I will not wasting my time to explore this anymore
<img width="953" alt="image" src="https://github.com/user-attachments/assets/f53e2e12-cbd4-4a8c-8b90-4f7559e10fbb">

**Reply by Yulongggggg** [reply]
@GiudGiud

**Reply by GiudGiud** [reply]
You can do that with the Nodal points. You just need to create a new NodalPositions object
for material properties that might give interesting results though, they are not defined on nodes in general.
